### PR TITLE
CP-2302- Notification Support

### PR DESCRIPTION
**Pull request by Azharhussain20**
Added a notification and user can retrieve the notification from local storage and api too.
file changes.
`CleverPush.h` & `CleverPush.m`
added methods:
`+ (NSArray*)getNotifications:(BOOL)combineWithApi` public method added on both `CleverPush.h/m`
`+ (NSString *)createDynamicURLForNotification`
`+ (NSMutableArray*)urlQueryParameter`
`+ (NSMutableArray *)getRemoteNotifications:(NSString*)url`

### CODE REVIEW

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 1]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+        return baseURL;
+    }
+}
+
+#pragma mark - Appending the topicId's as a query parameter.
++ (NSMutableArray*)urlQueryParameter {
+    NSMutableArray* subscriptionTopics = [self getSubscriptionTopics];
+    NSMutableArray* dynamicQueryParameter = [NSMutableArray new];
+    [subscriptionTopics enumerateObjectsUsingBlock: ^(id topic, NSUInteger index, BOOL *stop) {
+        NSString *queryParameter = [NSString stringWithFormat: @"topics[]=%@&", topic];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1991]: Does it work with multiple topics[]?

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 2]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+        NSString* appendableQueryParameters = [dynamicQueryParameters componentsJoinedByString:@""];
+        NSString *concatenatedURL = [NSString stringWithFormat:@"%@%@", baseURL, appendableQueryParameters];
+        return concatenatedURL;
+    } else {
+        return baseURL;
+    }
+}
+
+#pragma mark - Appending the topicId's as a query parameter.
++ (NSMutableArray*)urlQueryParameter {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1987]: Please rename this method to something descriptive e.g. `getReceivedNotificationsQueryParameters`

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 3]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+    NSMutableArray* dynamicQueryParameter = [NSMutableArray new];
+    [subscriptionTopics enumerateObjectsUsingBlock: ^(id topic, NSUInteger index, BOOL *stop) {
+        NSString *queryParameter = [NSString stringWithFormat: @"topics[]=%@&", topic];
+        [dynamicQueryParameter addObject:queryParameter];
+    }];
+    return dynamicQueryParameter;
+}
+
+#pragma mark - Get the Notifications based on the topic dialog Id's.
++ (NSMutableArray *)getRemoteNotifications:(NSString*)url {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1998]: Please rename to getReceivedNotificationsFromApi

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 4]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+        NSMutableArray *remoteNotifications = [self getRemoteNotifications:combinedURL];
+        return [combinedArray arrayByAddingObjectsFromArray:remoteNotifications];
+    } else {
+        return [self getNotifications];
+    }
+}
+
+#pragma mark - Creating URL based on the topic dialogue and append the topicId's as a query parameter.
++ (NSString *)createDynamicURLForNotification {
+    NSString *baseURL = [NSString stringWithFormat:@"https://api.cleverpush.com/channel/%@/received-notifications?", channelId];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1975]: Why do we need baseURL? The `path` in `[CleverPushHTTPClient sharedClient]` does not need domain, protocol, it only needs path.

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 5]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+        NSString *queryParameter = [NSString stringWithFormat: @"topics[]=%@&", topic];
+        [dynamicQueryParameter addObject:queryParameter];
+    }];
+    return dynamicQueryParameter;
+}
+
+#pragma mark - Get the Notifications based on the topic dialog Id's.
++ (NSMutableArray *)getRemoteNotifications:(NSString*)url {
+    __block NSMutableArray* notiications = nil;
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:2000]: Please do not work with semaphores here. The `getNotifications` method can have a callback block if needed.

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 6]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
     return notifications;
 }
 
+#pragma mark - Retrieving notifications based on the flag remote/local
++ (NSArray*)getNotifications:(BOOL)combineWithApi {
+    if (combineWithApi) {
+        NSString *combinedURL = [self createDynamicURLForNotification];
+        NSMutableArray* combinedArray = [[self getNotifications] mutableCopy];
+        NSMutableArray *remoteNotifications = [self getRemoteNotifications:combinedURL];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1966]: We need to merge the results without duplicates, only include `remoteNotifications` which are *not* included in `localNotifications`
Also make sure they are sorted correctly.

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 1 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:1991]: Yes It will work.

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 7]
```diff
@@ -1958,6 +1958,60 @@ + (NSArray*)getNotifications {
...
+
+#pragma mark - Get the Notifications based on the topic dialog Id's.
++ (NSMutableArray *)getRemoteNotifications:(NSString*)url {
+    __block NSMutableArray* notiications = nil;
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    
+    NSMutableURLRequest* request = [[CleverPushHTTPClient sharedClient] requestWithMethod:@"GET" path:url];
+    [self enqueueRequest:request onSuccess:^(NSDictionary* result) {
+        if (result != nil) {
+            notiications = [result valueForKey:@"notifications"];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:2005]: typo

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 6 above)
Comment (gbhrdt) [reply CleverPush/CleverPush/Source/CleverPush.m:1966]: When you build the array to return, please call `CPNotification initWithJson` for each entry

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 6 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:1966]: We didn't use any custom object yet to pass the data, We have always passed the data by NSArray and to link with cross platform like flutter we will need a Native objects.

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 8]
```diff
@@ -1994,22 +2008,52 @@ + (NSMutableArray*)urlQueryParameter {
     return dynamicQueryParameter;
 }
 
+#pragma mark - Remove duplicate entries from array.
++ (NSArray *)removeDuplicateEntriesFromArray:(NSArray *)array basedOnKey:(NSString *)key {
+    NSMutableArray *newArray = [NSMutableArray new];
+    NSArray *keysArray = [array valueForKey:key];
+    NSSet *noDuplicateKeys = [[NSSet alloc]initWithArray:keysArray];
+    for (NSString *currentKey in noDuplicateKeys) {
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@",key ,currentKey];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:2017]: Please fix comma syntax `,key ,currentKey`

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 6 above)
Comment (gbhrdt) [reply CleverPush/CleverPush/Source/CleverPush.m]: @Azharhussain20  Can you please check tomorrow that Flutter still works with this new method? As we now use CPNotification.

Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m]: Ok sure.

**Pull request merged**